### ISSUE TITLE
Resampled audio sounds like noise

### ISSUE THREAD

**Issue by izzytwosheds** [google/oboe#1482]
I am trying to integrate oboe resampler into my video transcoder project to be able to resample audio track: https://github.com/linkedin/LiTr/pull/160
Experiencing a strange bug: resampling by integer factor (44100 to 22050) using `Fastest` works fine, but different factor (44100 to 48000) or using any other resampling method makes resampled track sound like noise. Input values are produced by `MediaCodec` and are represented as `short` (from -32768 to 32767), normalizing them to (-1.0 to 1.0) does not help. I still have a suspicion that I am not using oboe resampler the way it is expected to be used, asking for guidance. Thank you!

**Comment by philburk** [comment]
I looked at audio-resampler.cpp
Your calls to the resampler look OK. But I see problems in data type conversion.
But that does not explain why some rates work and some don't.
Does downsampling work but upsampling make noise?

jbyte is signed. So the low bytes will get sign extended, which corrupts the higher bytes.
Either mask the signed bytes with 0xFF or make the sourceBuffer a (uint8_t *).

See [other comments](https://github.com/linkedin/LiTr/pull/160/files#r790192379) as well.

**Comment by izzytwosheds** [comment]
Masking with 0xFF or casting to (uint8_t *) didn't help. I agree it doesn't explain distorted output, subsampled tracks sound fine. I will add masking regardless, it is a right thing to do. Thank you for catching that! 

I also tried experimenting with passing `jshort` (one of my older working iterations did that) to compose `Short` in Java, and then cast it to `float` in JNI. That didn't work, either.

Integer subsampling (`Fastest` resampling to 22050, 11025) works fine, everything else fails (I presume, because it involves interpolation):
 - any super/sub sampling with other than `Fastest` resampling method
 - any super/sub sampling with non-integer scale, even with `Fastest` method

**Comment by izzytwosheds** [comment]
I figured it out. The problem was in multiple conversions between jbyte, short and float, and back to jbyte. Somewhere in the process little endian values were converted into big endian ones, which expectedly got rendered as a loud noise. I simplified my logic to allocate a direct buffer in Java and works with its bytes directly in JNI. Works like a charm - audio sounds great and I am not allocating and releasing ton of buffers. https://github.com/linkedin/LiTr/pull/160/files#diff-b481d04538d2ba92b6884b04636c6b61241412b43d7ed17a03e29c49d691f1dbR32

Thank you for amazing library! Audio sounds great, integrating oboe was well worth the effort.
